nikola/utils.py

```diff
@@ -2028,3 +2028,111 @@ def read_from_config(self, site, basename, posts_per_classification_per_language
+    else:
+        reStr = _DEFAULT_REST_METADATA_PARSING
+    result = reStr.findall(line.strip())
+    if match and result:
+        return (match, result[0])
+    elif not match and result:
+        return (result[0][0], result[0][1].strip())
+    else:
+        return (None,)
+
+
+def extract_metadata(file_lines):
+    """Extract metadata from the lines of a file.
+
+    Returns a pair ``(meta, metadata_type)``, where ``meta`` is the
+    metadata dictionary and ``metadata_type`` the metadata format.
+
+    Valid values for ``metadata_type`` are:
+    * ``'none'``: no metadata was found (file was empty)
+    * ``'yaml'``: metadata in YAML format
+    * ``'toml'``: metadata in TOML format
+    * ``'rest'``: metadata in reST format (the standard Nikola
+                  reST-like metadata format)
+    """
+    meta = {}
+    if not file_lines:
+        return meta, 'none'
+
+    # Skip up to one empty line at the beginning (for txt2tags)
+    if not file_lines[0]:
+        file_lines = file_lines[1:]
+
+    # If 1st line is '---', then it's YAML metadata
+    if file_lines[0] == '---':
+        if yaml is None:
+            req_missing('pyyaml', 'use YAML metadata', optional=True)
+            raise ValueError('Error parsing metadata')
+        idx = file_lines.index('---', 1)
+        meta = yaml.safe_load('\n'.join(file_lines[1:idx]))
+        # We expect empty metadata to be '', not None
+        for k in meta:
+            if meta[k] is None:
+                meta[k] = ''
+        return meta, 'yaml'
+
+    # If 1st line is '+++', then it's TOML metadata
+    if file_lines[0] == '+++':
+        if toml is None:
+            req_missing('toml', 'use TOML metadata', optional=True)
+            raise ValueError('Error parsing metadata')
+        idx = file_lines.index('+++', 1)
+        meta = toml.loads('\n'.join(file_lines[1:idx]))
+        return meta, 'toml'
+
+    # First, get metadata from the beginning of the file,
+    # up to first empty line
+
+    for line in file_lines:
+        if not line:
+            break
+        match = re_meta(line)
+        if match[0]:
+            meta[match[0]] = match[1]
+
+    return meta, 'nikola'
+
+
+def split_metadata(data):
+    """Split data from metadata in the raw post content.
+
+    This splits in the first empty line that is NOT at the beginning
+    of the document, or after YAML/TOML metadata without an empty line.
+
+    Returns a tuple ``(meta, content, metadata_type)`` where ``meta`` and
+    ``content`` are parts of ``data``, and ``metadata_type`` is the metadata
+    format.
+
+    Valid values for ``metadata_type`` are:
+    * ``'none'``: no metadata was found (file was empty)
+    * ``'yaml'``: metadata in YAML format
+    * ``'toml'``: metadata in TOML format
+    * ``'rest'``: metadata in reST format (the standard Nikola
+                  reST-like metadata format)
+    """
+    if data.startswith('---'):  # YAML metadata
```

Comment:
There should be something to split metadata based on the post object as well, and the post object should know what metadata type it was.